azurerm/resource_arm_recovery_services_protected_vm.go

```diff
@@ -117,11 +117,17 @@ func resourceArmRecoveryServicesProtectedVmCreateUpdate(d *schema.ResourceData,
 	if _, err = client.CreateOrUpdate(ctx, vaultName, resourceGroup, "Azure", containerName, protectedItemName, item); err != nil {
 		return fmt.Errorf("Error creating/updating Recovery Service Protected VM %q (Resource Group %q): %+v", protectedItemName, resourceGroup, err)
 	}
+	// Due to issue reported in PR https://github.com/terraform-providers/terraform-provider-azurerm/pull/3822
+	// Azure API, at the time, returned invalid state of the resource when backup_policy_id is updated
+	// Adding a graceful 15 seconds delay prior to the next call to Azure management seem to allow the API to correct itself
+	// and report back the correct status
+	time.Sleep(15 * time.Second)
```

Comment:
While this works, sleeping can be fraught with uncertainty as its hard to know exactly how long to wait. We should use a `WaitForState` call here instead.

Reply:
I have noted this block below in the method and the trace log in the previous comment was generated with the method `resourceArmRecoveryServicesProtectedVmWaitForState` executed which I believed `WaitForState` was also part of the method call.

The second part of the trace log in previous comment was from the method `resourceArmRecoveryServicesProtectedVmRead` yet it still showed incorrect status.

I have also observed a GET call, which is from `resourceArmRecoveryServicesProtectedVmWaitForState`, prior to state read and yet to see anything wrong with it hence my assumption was around a bug in Azure API or some strange symptom on how it responses to subsequent API request in a very short interval, perhaps with this particular type of resource. I may try to find another way without asking Azure to fix the issue but quite doubtful if it will work without adding a delay.

```
	if _, err = client.CreateOrUpdate(ctx, vaultName, resourceGroup, "Azure", containerName, protectedItemName, item); err != nil {
		return fmt.Errorf("Error creating/updating Recovery Service Protected VM %q (Resource Group %q): %+v", protectedItemName, resourceGroup, err)
	}
	// Due to issue reported in PR https://github.com/terraform-providers/terraform-provider-azurerm/pull/3822
	// Azure API, at the time, returned invalid state of the resource when backup_policy_id is updated
	// Adding a graceful 15 seconds delay prior to the next call to Azure management seem to allow the API to correct itself
	// and report back the correct status
	time.Sleep(15 * time.Second)

	resp, err := resourceArmRecoveryServicesProtectedVmWaitForState(client, ctx, true, vaultName, resourceGroup, containerName, protectedItemName)
	if err != nil {
		return err
	}

	id := strings.Replace(*resp.ID, "Subscriptions", "subscriptions", 1)
	d.SetId(id)

	return resourceArmRecoveryServicesProtectedVmRead(d, meta)

```

Reply:
I guess i understand a bit more about this when I had another quick look at the results
When updating `backup_policy_id`, it does not recreate any new resource so `resourceArmRecoveryServicesProtectedVmWaitForState` seems to return very fast with `Found` as the resource already exist but the response from Azure API is not fast enough to refresh the status from its actual backend data so it returned back with old attributes of the resource. I am looking to have another version of `resourceArmRecoveryServicesProtectedVmWaitForState` dedicated to update operation so we can compare the backup_policy_id before and after the request sent to Azure API. There should be only one attribute of the resource to compare on the update operation at this point given that only `backup_policy_id` is not `ForceNew` on change. I will submit another change for your review when I get back to update the code maybe in a day or so.

Reply:
Hi @katbyte 

I have made change to use `StateChangeConf` instead of arbitrary wait
It is on the same shared function `resourceArmRecoveryServicesProtectedVmWaitForState` still and the behaviour is driven by additional parameters that support all Create,Update,Delete operations. I believe that the change should have no effect to Create operation but I ran a test against the whole `TestAccAzureRMRecoveryServicesProtectedVm` module regardless.

Please have another review